### PR TITLE
Advanced network configuration

### PR DESCRIPTION
**Pull request by xenolog**
how it works:
- In site.pp early as it possible we call `prepare_network_config(Hash)` function. Function returns nothing. It has one incoming parameter -- hash with network configuration (interfaces, endpoints, transformations, roles). This function do nothing, only prepared and sanitized incoming parameters.
- create class for network configuration. Function generate_network_config() returns string with order network resources creation, only for debug purposes. Ex:

```
  class advanced_node_netconfig {
      $sdn = generate_network_config()
  }
```
- Call this class (may be in early stabe):

```
class {'advanced_node_netconfig':   stage => 'netconfig'  }
```

### CODE REVIEW

**Comment by xenolog** [comment]
After merging [PRD-1671] may be closed as completed.

**Comment by aglarendil** [comment]
+1

File: deployment/puppet/l23network/lib/puppet/parser/functions/generate_network_config.rb [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+    # Check for incorrect parameters
+    if not rv[:name].is_a? String
+      raise(Puppet::ParseError, "Unnamed transformation: '#{action}'.")
+    end
+    name = rv[:name]
+    if not rv[:bridge].is_a? String and not ["add-patch", "add-br"].index(action)
+      raise(Puppet::ParseError, "Undefined bridge for transformation '#{action}' with name '#{name}'.")
+    end
+    if action == "add-patch"
+      if not rv[:bridges].is_a? Array  and  rv[:bridges].size() != 2
```
Comment (warpc) [inline deployment/puppet/l23network/lib/puppet/parser/functions/generate_network_config.rb]: Please do not use `and` or `or`. Use `&&` or `||` instead.

File: deployment/puppet/l23network/lib/puppet/parser/functions/generate_network_config.rb [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        raise(Puppet::ParseError, "Transformation patch have wrong 'bridges' parameter.")
+      end
+      name = "patch__#{rv[:bridges][0]}__#{rv[:bridges][1]}"
+      if not rv[:peers].is_a? Array  and  rv[:peers].size() != 2
+        raise(Puppet::ParseError, "Transformation patch '#{name}' have wrong 'peers' parameter.")
+      end
+      rv[:name] = name
+    end
+    if action == "add-bond"
+      if not rv[:interfaces].is_a? Array or rv[:interfaces].size() != 2
```
Comment (warpc) [inline deployment/puppet/l23network/lib/puppet/parser/functions/generate_network_config.rb]: Please see above

File: deployment/puppet/l23network/lib/puppet/parser/functions/generate_network_config.rb [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+        if k.to_sym() == :IP
+          if !(v.is_a?(Array) || ['none','dhcp',nil].include?(v))
+            raise(Puppet::ParseError, "generate_network_config(): IP field for endpoint '#{e_name}' must be array of IP addresses, 'dhcp' or 'none'.")
+          elsif ['none','dhcp',nil].include?(v)
+            endpoints[e_name][:IP].insert(-1, v ? v : 'none')
+          else
+            v.each do |ip|
+              begin
+                iip = IPAddr.new(ip)
+                endpoints[e_name][:IP].insert(-1, ip)
```
Comment (warpc) [inline deployment/puppet/l23network/lib/puppet/parser/functions/generate_network_config.rb:168]:
``` ruby
endpoints[e_name][:IP] << ip
```

File: deployment/puppet/l23network/lib/puppet/parser/functions/get_network_role_property.rb [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      raise(Puppet::ParseError, "get_network_role_property(...): invalid IP address for endpoint '#{interface}'.")
+  end
+
+  if ipaddr_cidr == nil
+    return nil
+  end
+
+  case mode
+    when 'CIDR'
+      return ipaddr_cidr
```
Comment (warpc) [inline deployment/puppet/l23network/lib/puppet/parser/functions/get_network_role_property.rb]: you can safely remove `return` here

File: deployment/puppet/l23network/lib/puppet/parser/functions/lib/hash_tools.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      if v.is_a? Hash
+        rv.insert(-1, self.sanitize_keys_in_hash(v))
+      elsif v.is_a? Array
+        rv.insert(-1, self.process_array4keys(v))
+      else
+        rv.insert(-1, v)
+      end
+    end
+    return rv
+  end
```
Comment (warpc) [inline deployment/puppet/l23network/lib/puppet/parser/functions/lib/hash_tools.rb:15]: Check this comment: https://github.com/Mirantis/fuel/pull/603/files#diff-1532e0852970547bc124ec140a940a26R27

File: deployment/puppet/l23network/lib/puppet/parser/functions/lib/hash_tools.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  def self.sanitize_bool_in_hash(hh)
+    rv = {}
+    hh.each do |k, v|
+      if v.is_a? String or v.is_a? Symbol
```
Comment (warpc) [inline deployment/puppet/l23network/lib/puppet/parser/functions/lib/hash_tools.rb]: See comment above (or -> ||)

File: deployment/puppet/l23network/lib/puppet/parser/functions/lib/hash_tools.rb [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        rv[k] = case v.upcase()
+          when 'TRUE', :TRUE then true
+          when 'FALSE', :FALSE then false
+          when 'NONE', :NONE, 'NULL', :NULL, 'NIL', :NIL, 'NILL', :NILL then nil
+          else v
+        end
```
Comment (warpc) [inline deployment/puppet/l23network/lib/puppet/parser/functions/lib/hash_tools.rb:54]: ?))

File: deployment/puppet/l23network/lib/puppet/provider/l2_ovs_patch/ovs.rb [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+      cmd.concat(['--', 'set', 'interface', name, 'type=patch'])
+      begin
+        vsctl(cmd)
+      rescue Puppet::ExecutionFailure => errmsg
+        raise Puppet::ExecutionFailure, "Can't create patch '#{name}':\n#{errmsg}"
+      end
+      i += 1
+    end
+    i = 0
+    for name in names
```
Comment (warpc) [inline deployment/puppet/l23network/lib/puppet/provider/l2_ovs_patch/ovs.rb:68]:
``` ruby
names.each do |name|

end
```

File: deployment/puppet/l23network/lib/puppet/provider/l2_ovs_patch/ovs.rb [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+      begin
+        vsctl('del-port', @resource[:bridges][i], name)
+      rescue Puppet::ExecutionFailure => error
+        raise Puppet::ExecutionFailure, "Can't remove patch '#{name}' from bridge '#{@resource[:bridges][i]}':\n#{error}"
+      end
+      i += 1
+    end
+  end
```
Comment (warpc) [inline deployment/puppet/l23network/lib/puppet/provider/l2_ovs_patch/ovs.rb:89]:
``` ruby
def destroy
  names = get_names
  names.each_with_index do |name, i|
    begin
       vsctl('del-port', @resource[:bridges][i], name)
    rescue
       raise Puppet::ExecutionFailure, "Can't remove patch '#{name}' from bridge '#{@resource[:bridges][i]}':\n#{error}"
    end
end
```

File: deployment/puppet/l23network/lib/puppet/type/l2_ovs_bond.rb [hunk 10]
```diff
@@ -50,6 +50,47 @@
...
+        if not (val.is_a?(Array) or val.is_a?(String) or val.is_a?(Integer)) # String need for array with one element. it's a puppet's feature
+          fail("trunks must be an array (not #{val.class}).")
+        end
+      end
+
+      munge do |val|
+        if val.is_a?(String)
+          [val.to_i]
+        elsif val.is_a?(Integer)
+          if val >= 0 and val < 4095
```
Comment (warpc) [inline deployment/puppet/l23network/lib/puppet/type/l2_ovs_bond.rb:83]: and -> &&

File: deployment/puppet/l23network/lib/puppet/parser/functions/generate_network_config.rb (hunk 3 above)
Comment (xenolog) [inline deployment/puppet/l23network/lib/puppet/parser/functions/generate_network_config.rb:168]: not intuitive sintax

File: deployment/puppet/l23network/lib/puppet/parser/functions/lib/hash_tools.rb (hunk 7 above)
Comment (xenolog) [inline deployment/puppet/l23network/lib/puppet/parser/functions/lib/hash_tools.rb:54]: convert string representation bool value to bool type

**Comment by aglarendil** [comment]
+1

**Pull request merged**